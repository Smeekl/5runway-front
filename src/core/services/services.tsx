import React from 'react';
import { IServices } from '@serviceTypes';

const { createContext, useContext } = React;
const ServicesContext = createContext({} as IServices);

type TServiceProviderProps = {
  services: IServices;
  children: React.ReactNode;
};

export const ServiceProvider = ({
  services,
  children,
}: TServiceProviderProps) => (
  <ServicesContext.Provider value={services}>
    {children}
  </ServicesContext.Provider>
);

export const useService = () => useContext(ServicesContext);
