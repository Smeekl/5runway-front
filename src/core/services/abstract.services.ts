import { AxiosInstance } from 'axios';
import { StorageService } from './storage-services/storage.services';
import { TTokens } from './types';

export class AbstractApiClient {
  api: AxiosInstance;
  storageService: StorageService;

  constructor(api: AxiosInstance, storageService: StorageService) {
    this.api = api;
    this.storageService = storageService;
  }

  setTokens({ accessToken, refreshToken }: TTokens) {
    this.storageService.set('accessToken', accessToken);
    this.storageService.set('refreshToken', refreshToken);
  }
}
