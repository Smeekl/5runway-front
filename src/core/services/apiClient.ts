import Axios, { AxiosInstance } from 'axios';
import { StorageService } from './storage-services/storage.services';

export function CreateApiClient(
  baseURL: string,
  storageService: StorageService
): AxiosInstance {
  const api = Axios.create({ baseURL });

  let retry = false;

  api.interceptors.request.use(
    (config) => {
      if (!config.headers) {
        config.headers = {};
      }

      config.headers.Authorization = `Bearer ${storageService.get(
        'accessToken'
      )}`;

      return config;
    },
    (error) => Promise.reject(error)
  );
  api.interceptors.response.use(
    (res) => res,
    async (err) => {
      if (err.response.status === 401 && !retry) {
        retry = true;

        const refreshToken = storageService.get('refreshToken');

        if (!refreshToken) {
          return;
        }

        try {
          const { data } = await api.post('/auth/refresh', { refreshToken });

          storageService.set('accessToken', data.accessToken);
          storageService.set('refreshToken', data.refreshToken);
        } catch (error) {
          throw new Error('Token expired');
        }
        return;
      }

      retry = false;

      return Promise.reject(err);
    }
  );

  return api;
}
