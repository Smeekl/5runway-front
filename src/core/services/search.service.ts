import { AxiosInstance } from 'axios';
import { AbstractApiClient } from './abstract.services';
import { StorageService } from './storage-services/storage.services';

export class SearchService extends AbstractApiClient {
  dataStorage: StorageService;

  constructor(api: AxiosInstance, dataStorage: StorageService) {
    super(api);
    this.dataStorage = dataStorage;
  }

  searchItem = async (value: string) => {
    this.dataStorage.set(
      'searchedItem',
      JSON.stringify([value, ...this.getSearchedItems()].slice(0, 10))
    );

    try {
      const res = await this.api.get(value);

      return res.data;
    } catch (error) {}
  };

  getSearchedItems = () =>
    JSON.parse(this.dataStorage.get('searchedItem') || '[]');

  clearSearchedItems = () => this.dataStorage.clear();
}
