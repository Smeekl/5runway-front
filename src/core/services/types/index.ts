import { AuthService } from '@services/auth.service';
import { SearchService } from '@services/search.service';
import { StorageService } from '@services/storage-services/storage.services';
import { UserService } from '@services/user.service';

export interface IServices {
  storageService: StorageService;
  userService: UserService;
  searchService: SearchService;
  authService: AuthService;
}

export type TTokens = {
  accessToken: string;
  refreshToken: string;
};

export type TFilterOptions = {
  model: string;
  location: string;
  minPrice: number;
  maxPrice: number;
  minMileage: number;
  maxMileage: number;
};
