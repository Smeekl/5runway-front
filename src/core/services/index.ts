import { CreateApiClient } from './apiClient';
import { StorageService } from './storage-services/storage.services';
import { UserService } from './user.service';
import { SearchService } from './search.service';
import { AuthService } from './auth.service';

export function createServices() {
  const storageService = new StorageService(window.localStorage);
  const api = CreateApiClient(process.env.REACT_APP_BASEURL, storageService);
  const searchService = new SearchService(api, storageService);
  const userService = new UserService(api, storageService);
  const authService = new AuthService(api, storageService);

  return {
    storageService,
    searchService,
    userService,
    authService,
  };
}
