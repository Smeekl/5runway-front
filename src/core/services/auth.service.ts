import { AbstractApiClient } from './abstract.services';
import { TTokens } from './types';

export class AuthService extends AbstractApiClient {
  googleAuth = async (google_token: string) => {
    const { data } = await this.api.get<TTokens>(`/auth/google/verification?token=${google_token}`);

    this.setTokens(data);
  }
}
