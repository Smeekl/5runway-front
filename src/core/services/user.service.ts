import { TUser } from '@composed-components/types';
import { AbstractApiClient } from './abstract.services';
import { ENDPOINTS } from '@constants';

export class UserService extends AbstractApiClient {
  registerUser({ email, password }) {
    return this.api.post('/auth/registration', { email, password });
  }

  async loginUser(email: string, password: string) {
    const { data } = await this.api.post('/auth/login', { email, password });

    this.setTokens(data);

    const user = await this.api.get('/user/profile');

    this.storageService.set('user', JSON.stringify(user.data));
  }

  getUser() {
    return this.api.get<TUser>('/user/profile');
  }

  logoutUser() {
    this.storageService.clear();
  }

  restorePassword(email: string) {
    return this.api.post(ENDPOINTS.RESTORE_PASSWORD, { email });
  }

  resetPassword(token: string, password: string) {
    return this.api.post(ENDPOINTS.RESET_PASSWORD, { token, password });
  }
}
