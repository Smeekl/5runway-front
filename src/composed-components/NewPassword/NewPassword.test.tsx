import { MemoryRouter } from 'react-router-dom';
import { render } from '@testing-library/react';
import { waitFor } from '@testing-library/dom';
import { createMemoryHistory } from 'history';
import { ServiceProvider } from '@services/services';
import NewPasswordSidebar from './NewPassword';
import { createServices } from '@services/index';
import LocalStorageMock from '@mocks/localStorage.mock';

describe('NewPasswordSidebar UI tests', () => {
  test('should render correctly', () => {
    const services = createServices();

    const { container } = render(
      <MemoryRouter>
        <ServiceProvider services={services}>
          <NewPasswordSidebar />
        </ServiceProvider>
      </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
  });

  test('should be in the document', () => {
    const services = createServices();

    const { container } = render(
      <MemoryRouter>
        <ServiceProvider services={services}>
          <NewPasswordSidebar />
        </ServiceProvider>
      </MemoryRouter>
    );
    expect(container).toBeInTheDocument();
  });

  test('should add temp token in localStorage', () => {
    const services = createServices();
    const localStorageMock = new LocalStorageMock();

    render(
      <MemoryRouter>
        <ServiceProvider services={services}>
          <NewPasswordSidebar />
        </ServiceProvider>
      </MemoryRouter>
    );
    expect(localStorageMock.getItem('token')).toBeTruthy();
  });

  test('should redirect to url without token', async () => {
    const services = createServices();
    const history = createMemoryHistory({
      initialEntries: [
        '/login/confirm-new-password/?token=12412digjeogijsdfgs.4osdjgiauj9-428u3opiqjefpoasdfa.sdjgapdgjioj23pio4h12',
      ],
    });

    render(
      <MemoryRouter>
        <ServiceProvider services={services}>
          <NewPasswordSidebar />
        </ServiceProvider>
      </MemoryRouter>
    );

    await waitFor(() => {
      expect(history.location.pathname).toEqual('/login/confirm-new-password/');
    });
  });
});
