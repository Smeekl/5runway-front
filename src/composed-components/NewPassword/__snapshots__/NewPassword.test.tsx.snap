// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewPasswordSidebar UI tests should render correctly 1`] = `
<div>
  <div
    class="new-password"
  >
    <h3
      class="new-password__title"
    >
      New Password
    </h3>
    <form
      action="#"
    >
      <p
        class="new-password__name"
      >
        New Password
      </p>
      <div
        class="sidebar__field-wrapper"
      >
        <input
          class="new-password__field"
          name="password"
          type="password"
        />
        <span
          aria-label="eye"
          class="anticon anticon-eye sidebar__eye-icon"
          role="img"
          tabindex="-1"
        >
          <svg
            aria-hidden="true"
            data-icon="eye"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M942.2 486.2C847.4 286.5 704.1 186 512 186c-192.2 0-335.4 100.5-430.2 300.3a60.3 60.3 0 000 51.5C176.6 737.5 319.9 838 512 838c192.2 0 335.4-100.5 430.2-300.3 7.7-16.2 7.7-35 0-51.5zM512 766c-161.3 0-279.4-81.8-362.7-254C232.6 339.8 350.7 258 512 258c161.3 0 279.4 81.8 362.7 254C791.5 684.2 673.4 766 512 766zm-4-430c-97.2 0-176 78.8-176 176s78.8 176 176 176 176-78.8 176-176-78.8-176-176-176zm0 288c-61.9 0-112-50.1-112-112s50.1-112 112-112 112 50.1 112 112-50.1 112-112 112z"
            />
          </svg>
        </span>
      </div>
      <div
        class="sidebar__error-indicator "
      >
        <span
          aria-label="exclamation-circle"
          class="anticon anticon-exclamation-circle sidebar__error-icon"
          role="img"
        >
          <svg
            aria-hidden="true"
            data-icon="exclamation-circle"
            fill="currentColor"
            focusable="false"
            height="1em"
            viewBox="64 64 896 896"
            width="1em"
          >
            <path
              d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
            />
            <path
              d="M464 688a48 48 0 1096 0 48 48 0 10-96 0zm24-112h48c4.4 0 8-3.6 8-8V296c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v272c0 4.4 3.6 8 8 8z"
            />
          </svg>
        </span>
      </div>
      <div
        class="new-password__block"
      >
        <p
          class="new-password__name"
        >
          Confirm Password
        </p>
        <input
          class="new-password__field"
          name="confirmPassword"
          type="password"
          value=""
        />
        <div
          class="sidebar__error-indicator "
        >
          <span
            aria-label="exclamation-circle"
            class="anticon anticon-exclamation-circle sidebar__error-icon"
            role="img"
          >
            <svg
              aria-hidden="true"
              data-icon="exclamation-circle"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
              />
              <path
                d="M464 688a48 48 0 1096 0 48 48 0 10-96 0zm24-112h48c4.4 0 8-3.6 8-8V296c0-4.4-3.6-8-8-8h-48c-4.4 0-8 3.6-8 8v272c0 4.4 3.6 8 8 8z"
              />
            </svg>
          </span>
        </div>
      </div>
      <button
        class="ant-btn ant-btn-default new-password__btn"
        disabled=""
        type="submit"
      >
        <span>
          Save
        </span>
      </button>
    </form>
  </div>
</div>
`;
